### ISSUE TITLE
Returning both successes and failures

### ISSUE THREAD

**Issue by nasht00** [CacheControl/json-rules-engine#56]
The README says that "// run() returns events with truthy conditions" in this example:

```
// Run the engine to evaluate
engine
  .run(facts)
  .then(events => { // run() returns events with truthy conditions
    events.map(event => console.log(event.params.message))
  })
```

What about when the result is false?

What I am trying to achieve is something like:
```
// Run the engine to evaluate
engine
  .run(facts)
  .then(events => { 
    return true // return false
  })
```

I want my function to return true if the conditions are met, false if they are not.
What's the way to achieve this?

**Comment by alank64** [comment]
I think you want to use the

```
engine.on('success'...)
engine.on('failure'...)
```

So that on a failed event, you can return a false... that's if I understood your question correctly?

**Comment by CacheControl** [comment]
@nasht00 I believe the `on.('failure')` hook is indeed what you're looking for.  See [docs](https://github.com/CacheControl/json-rules-engine/blob/master/docs/engine.md#engineonfailure-functionobject-event-almanac-almanac-ruleresult-ruleresult) for the spec and also [this example](https://github.com/CacheControl/json-rules-engine/blob/master/examples/09-rule-results.js).  Thanks @alank64 for the quick response.

**Comment by nasht00** [comment]
@alank64 I'm not clear about `res.data.events.failed.push(event);`.
What is that?

What I am trying to achieve is pretty simple:

I am writing a method called `checkEligibility(facts)` which should return a boolean (or a Promise of boolean if necessary).

The `run` method only returns success events.

My current workaround is like this:

```
    // run the engine
    return engine
        .run(facts)
        .then(function (triggeredEvents) { // engine returns a list of events with truthy conditions
            if (triggeredEvents.length > 0) { //there was a success
                return Promise.resolve(true);
            } else {
                return Promise.resolve(false);
            }
        });
```

I am basically assuming that if the `triggeredEvents` array is empty, it must mean that the result should be `false`.
Not sure if that looks like a best practice ...

**Comment by knalbandianbrightgrove** [comment]
@nasht00 

Try this one

`
// run the engine
return engine
    .run(facts)
    .then(function (triggeredEvents) { // engine returns a list of events with truthy conditions
        if (triggeredEvents.length > 0) { //there was a success
            return Promise.resolve(true);
        }
    }, function () {
       return Promise.resolve(false);
    });
`

**Comment by nasht00** [comment]
@knalbandianbrightgrove can you fix the markdown syntax, I can't read the code snippet ...

**Comment by nasht00** [comment]
Thanks @knalbandianbrightgrove , I think it's pretty much the same algorithm that I posted above no?
Except you are also handling the case where `run` fails completely?

**Comment by nasht00** [comment]
I understand. Yet in my tests, when the rule is false, it still reached the `then` part, just with empty events array.

My assumption is that it reaches the failure callback only if something goes seriously wrong during the `run` (still a good thing to catch indeed), and not simply when no success event was triggered.
That's what I understood from the documentation (https://github.com/CacheControl/json-rules-engine/blob/master/docs/engine.md#enginerunobject-facts-object-options---promise-events):

> Returns a promise which resolves when all rules have been run.

**Comment by alank64** [comment]
@nasht00 sorry, I used a part of my code here, in that I'm simply pushing the failed or success events into an array (res.data.events.failed or res.data.events.success) to return to the API caller, add them to what ever your callback or response is.

The idea is that any failed events or success are returned in the .on hook, use them to record event failures or successes. However you brought up a good question which I noticed the other day, and maybe I need to perform more testing, is that if one event fails does json-rules-engine continue on other events. Something I'm about to test now.
